File: edukate-backend/src/main/java/io/github/sanyavertolet/edukate/backend/services/CheckResultService.java
package io.github.sanyavertolet.edukate.backend.services;

import io.github.sanyavertolet.edukate.backend.entities.Submission;
import io.github.sanyavertolet.edukate.backend.repositories.CheckResultRepository;
import io.github.sanyavertolet.edukate.backend.entities.CheckResult;
import io.github.sanyavertolet.edukate.common.SubmissionStatus;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

@Service
@AllArgsConstructor
public class CheckResultService {
    private final CheckResultRepository checkResultRepository;
    private final SubmissionService submissionService;

    @Transactional
    public Mono<Tuple2<CheckResult, Submission>> saveAndUpdateSubmission(CheckResult checkResult) {
        return submissionService.findById(checkResult.getSubmissionId())
                .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND, "Submission not found")))
                .flatMap(submission -> checkResultRepository.save(checkResult)
                        .map(saved -> Tuples.of(saved, submission)))
                .flatMap(tuple -> {
                    CheckResult saved = tuple.getT1();
                    Submission submission = tuple.getT2();
                    SubmissionStatus newStatus = SubmissionStatus.from(saved.getStatus());
                    submission.setStatus(SubmissionStatus.best(submission.getStatus(), newStatus));
                    return submissionService.update(submission).thenReturn(tuple);
                });
    }

    public Mono<CheckResult> findById(String id) {
        return checkResultRepository.findById(id);
    }

    public Flux<CheckResult> findAllBySubmissionId(String submissionId) {
        return checkResultRepository.findBySubmissionId(
                submissionId, Sort.by(Sort.Direction.DESC, "createdAt")
        );
    }
}

File: edukate-backend/src/main/kotlin/io/github/sanyavertolet/edukate/backend/services/CheckResultService.kt
package io.github.sanyavertolet.edukate.backend.services

import io.github.sanyavertolet.edukate.backend.entities.CheckResult
import io.github.sanyavertolet.edukate.backend.repositories.CheckResultRepository
import io.github.sanyavertolet.edukate.common.checks.CheckResultMessage
import io.github.sanyavertolet.edukate.common.utils.orNotFound
import io.micrometer.core.instrument.MeterRegistry
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class CheckResultService(
    private val checkResultRepository: CheckResultRepository,
    private val meterRegistry: MeterRegistry,
) {
    /** Direct insert — used for self-check. The DB trigger handles submission status update. */
    fun saveCheckResult(checkResult: CheckResult): Mono<CheckResult> =
        checkResultRepository.save(checkResult).doOnNext {
            meterRegistry.counter("check.outcomes", "status", it.status.name).increment()
        }

    /**
     * Finds the stub by its ID and updates it in-place with the checker result. The DB trigger handles submission status +
     * problem_progress update automatically.
     */
    fun updateFromMessage(message: CheckResultMessage): Mono<CheckResult> =
        checkResultRepository
            .findById(message.checkResultId)
            .orNotFound("CheckResult ${message.checkResultId} not found")
            .map { stub ->
                stub.copy(
                    status = message.status,
                    trustLevel = message.trustLevel,
                    errorType = message.errorType,
                    explanation = message.explanation,
                )
            }
            .flatMap { checkResultRepository.save(it) }
            .doOnNext { meterRegistry.counter("check.outcomes", "status", it.status.name).increment() }

    fun findById(id: Long): Mono<CheckResult> = checkResultRepository.findById(id)

    fun findAllBySubmissionId(submissionId: Long): Flux<CheckResult> =
        checkResultRepository.findBySubmissionId(submissionId, Sort.by(Sort.Direction.DESC, "createdAt"))
}
